File: src/pages/rolling-blocks-solver/a-star/Replay.cpp
#include "Replay.h"

#include <algorithm>

namespace {

// A block counts towards goal coverage only if its WHOLE footprint is on goal
// tiles — a block half on a goal contributes nothing.
bool footprintAllGoal(const replay::Puzzle &puzzle, const Block &block) {
  const uint8_t w = puzzle.gridWidth;
  for (int8_t cx = block.x; cx < block.x + static_cast<int8_t>(block.width);
       cx++) {
    for (int8_t cy = block.y; cy < block.y + static_cast<int8_t>(block.depth);
         cy++) {
      if (puzzle.cells[positionToIndex(cx, cy, w)] != Tile::Goal) {
        return false;
      }
    }
  }
  return true;
}

void markFootprint(const replay::Puzzle &puzzle, const Block &block,
                   boost::dynamic_bitset<> &covered) {
  const uint8_t w = puzzle.gridWidth;
  for (int8_t cx = block.x; cx < block.x + static_cast<int8_t>(block.width);
       cx++) {
    for (int8_t cy = block.y; cy < block.y + static_cast<int8_t>(block.depth);
         cy++) {
      covered.set(positionToIndex(cx, cy, w));
    }
  }
}

// Strict solved test: all must-touch cells satisfied AND the set of goal
// cells under fully-on-goal blocks is every goal cell.
bool isSolved(const replay::Puzzle &puzzle, const std::vector<Block> &blocks,
              const boost::dynamic_bitset<> &satisfied) {
  size_t goalCells = 0;
  for (size_t i = 0; i < puzzle.cells.size(); i++) {
    if (puzzle.cells[i] == Tile::MustTouch && !satisfied.test(i)) {
      return false;
    }
    if (puzzle.cells[i] == Tile::Goal) {
      goalCells++;
    }
  }
  if (goalCells == 0) {
    return true;
  }
  boost::dynamic_bitset covered(puzzle.cells.size());
  for (const auto &block : blocks) {
    if (footprintAllGoal(puzzle, block)) {
      markFootprint(puzzle, block, covered);
    }
  }
  return covered.count() == goalCells;
}

} // namespace

namespace replay {

Outcome replayTurns(const Puzzle &puzzle, const std::vector<Turn> &turns) {
  Outcome outcome;
  std::vector<Block> blocks = puzzle.blocks;
  boost::dynamic_bitset satisfied(puzzle.cells.size());
  for (const auto &block : blocks) {
    satisfied =
        block.updateMustTouchCells(puzzle.gridWidth, puzzle.cells, satisfied);
  }

  if (isSolved(puzzle, blocks, satisfied)) {
    outcome.firstSolvedAt = 0;
  }

  for (size_t i = 0; i < turns.size(); i++) {
    const auto it = std::ranges::find_if(blocks, [&](const Block &b) {
      return b.id == turns[i].blockId;
    });
    if (it == blocks.end()) {
      return outcome; // legal stays false
    }
    it->roll(turns[i].direction);
    if (!it->checkValidity(puzzle.gridWidth, puzzle.gridHeight, puzzle.cells,
                           blocks, satisfied)) {
      return outcome;
    }
    satisfied =
        it->updateMustTouchCells(puzzle.gridWidth, puzzle.cells, satisfied);
    if (outcome.firstSolvedAt == SIZE_MAX &&
        isSolved(puzzle, blocks, satisfied)) {
      outcome.firstSolvedAt = i + 1;
    }
  }

  outcome.legal = true;
  outcome.solvedAtEnd = isSolved(puzzle, blocks, satisfied);
  return outcome;
}

bool applyTurns(const Puzzle &puzzle, const std::vector<Turn> &turns,
                std::vector<Block> &blocksOut,
                boost::dynamic_bitset<> &satisfiedOut) {
  blocksOut = puzzle.blocks;
  satisfiedOut.resize(puzzle.cells.size());
  satisfiedOut.reset();
  for (const auto &block : blocksOut) {
    satisfiedOut = block.updateMustTouchCells(puzzle.gridWidth, puzzle.cells,
                                              satisfiedOut);
  }
  for (const auto &[blockId, direction] : turns) {
    const auto it = std::ranges::find_if(blocksOut, [&](const Block &b) {
      return b.id == blockId;
    });
    if (it == blocksOut.end()) {
      return false;
    }
    it->roll(direction);
    if (!it->checkValidity(puzzle.gridWidth, puzzle.gridHeight, puzzle.cells,
                           blocksOut, satisfiedOut)) {
      return false;
    }
    satisfiedOut =
        it->updateMustTouchCells(puzzle.gridWidth, puzzle.cells, satisfiedOut);
  }
  return true;
}

std::vector<Turn> truncateToEarliestSolve(const Puzzle &puzzle,
                                          const std::vector<Turn> &turns) {
  const auto [legal, solvedAtEnd, firstSolvedAt] = replayTurns(puzzle, turns);
  if (!legal || firstSolvedAt == SIZE_MAX || firstSolvedAt >= turns.size()) {
    return turns;
  }
  return {turns.begin(), turns.begin() + static_cast<ptrdiff_t>(firstSolvedAt)};
}

} // namespace replay

File: src/pages/rolling-blocks-solver/replay.ts
import type { Tile } from "../../util/types";
import { extractBit, positionToIndex } from "../../util/utilMethods";
import type { Block } from "./block";
import type { Turn } from "./turn";

/**
 * Replays a rolling-blocks solution in TypeScript.
 *
 * This is a second copy of the semantics in `a-star/Replay.cpp`, and the two
 * must stay in sync — the C++ one is the oracle the solver and its tests use,
 * this one is what the page needs to draw the board between two moves. Every
 * rule it implements already lives in `Block`; the module only sequences them.
 */

export interface Puzzle {
  gridWidth: number;
  gridHeight: number;
  /** Column-major, exactly like `Board.cells`. */
  cells: Tile[][];
  blocks: Block[];
}

/** The board as it stands before one particular turn is played. */
export interface ReplayStep {
  blocks: Block[];
  /** Bitmask of must-touch cells already stepped on. */
  satisfied: bigint;
}

export interface ReplayResult {
  /**
   * `steps[i]` is the state turn `i` is played on, so a legal replay of `n`
   * turns yields `n + 1` entries and the last one is the final board.
   */
  steps: ReplayStep[];
  /** How many turns could actually be played; `turns.length` when all could. */
  legalUpTo: number;
}

function cloneAll(blocks: Block[]): Block[] {
  return blocks.map(block => block.clone());
}

/** Seeds the satisfied mask from where the blocks already stand. */
function initialSatisfied(puzzle: Puzzle, blocks: Block[]): bigint {
  let satisfied = 0n;
  for (const block of blocks) {
    satisfied = block.updateMustTouchCells(
      puzzle.gridWidth,
      puzzle.cells,
      satisfied,
    );
  }
  return satisfied;
}

/**
 * A block counts towards goal coverage only if its WHOLE footprint is on goal
 * tiles — a block half on a goal contributes nothing.
 */
function footprintAllGoal(puzzle: Puzzle, block: Block): boolean {
  for (let x = block.x; x < block.x + block.width; x++) {
    for (let y = block.y; y < block.y + block.depth; y++) {
      if (puzzle.cells[x]?.[y] !== "goal") return false;
    }
  }
  return true;
}

/** Whether every must-touch cell on the board has been stepped on. */
function allStarsCollected(puzzle: Puzzle, satisfied: bigint): boolean {
  const { gridWidth, gridHeight, cells } = puzzle;
  for (let x = 0; x < gridWidth; x++) {
    for (let y = 0; y < gridHeight; y++) {
      if (cells[x]?.[y] !== "mustTouch") continue;
      if (extractBit(satisfied, positionToIndex(x, y, gridWidth)) === 0n) {
        return false;
      }
    }
  }
  return true;
}

/** The `"x,y"` key of every goal cell on the board. */
function goalKeys(puzzle: Puzzle): Set<string> {
  const keys = new Set<string>();
  for (let x = 0; x < puzzle.gridWidth; x++) {
    for (let y = 0; y < puzzle.gridHeight; y++) {
      if (puzzle.cells[x]?.[y] === "goal") keys.add(`${x},${y}`);
    }
  }
  return keys;
}

/** Drops every cell under `block` from `keys`. */
function removeFootprint(block: Block, keys: Set<string>) {
  for (let x = block.x; x < block.x + block.width; x++) {
    for (let y = block.y; y < block.y + block.depth; y++) {
      keys.delete(`${x},${y}`);
    }
  }
}

/**
 * Strict solved test: every must-touch cell satisfied, and every goal cell
 * covered by a block that stands entirely on goal tiles. A board with no goal
 * cells passes the second half vacuously.
 */
export function isSolved(
  puzzle: Puzzle,
  blocks: Block[],
  satisfied: bigint,
): boolean {
  if (!allStarsCollected(puzzle, satisfied)) return false;

  const goalCells = goalKeys(puzzle);
  if (goalCells.size === 0) return true;

  for (const block of blocks) {
    if (footprintAllGoal(puzzle, block)) removeFootprint(block, goalCells);
  }
  return goalCells.size === 0;
}

/**
 * Whether the board is already solved where it stands, no turn played.
 *
 * Worth asking because the solver answers such a board with the EMPTY plan,
 * which is exactly what an arm that found nothing returns — so "no moves" has
 * to be told apart from "no solution" somewhere, and the start state is the
 * only thing that can say which it is.
 *
 * Goes through `replayTurns` rather than reading the blocks directly: seeding
 * the satisfied mask from where they already stand is a rule of the replay, and
 * a second copy of it here is a second place for it to go stale.
 */
export function isSolvedAtStart(puzzle: Puzzle): boolean {
  const start = replayTurns(puzzle, []).steps[0]!;
  return isSolved(puzzle, start.blocks, start.satisfied);
}

/**
 * Walks the turns, capturing the board before each one. Stops at the first
 * turn that names a missing block or leaves it in an illegal place — a witness
 * the page cannot play is worth showing up to the point it breaks, not
 * pretending past it.
 */
export function replayTurns(puzzle: Puzzle, turns: Turn[]): ReplayResult {
  let blocks = cloneAll(puzzle.blocks);
  let satisfied = initialSatisfied(puzzle, blocks);
  const steps: ReplayStep[] = [{ blocks: cloneAll(blocks), satisfied }];

  for (let i = 0; i < turns.length; i++) {
    const turn = turns[i]!;
    const next = cloneAll(blocks);
    const moving = next.find(block => block.id === turn.blockId);
    if (!moving) return { steps, legalUpTo: i };

    moving.roll(turn.direction);
    const legal = moving.checkValidity(
      puzzle.gridWidth,
      puzzle.gridHeight,
      puzzle.cells,
      next,
      satisfied,
    );
    if (!legal) return { steps, legalUpTo: i };

    satisfied = moving.updateMustTouchCells(
      puzzle.gridWidth,
      puzzle.cells,
      satisfied,
    );
    blocks = next;
    steps.push({ blocks: cloneAll(blocks), satisfied });
  }

  return { steps, legalUpTo: turns.length };
}
